runtime/drivers/duckdb/olap.go

```diff
@@ -181,6 +186,374 @@ func (c *connection) EstimateSize() (int64, bool) {
+	err = c.Exec(ctx, &drivers.Statement{
+		Query:    fmt.Sprintf("ALTER TABLE %s.default ALTER %s TYPE %s", safeSQLName(dbName), safeSQLName(columnName), newType),
+		Priority: 1,
+	})
+	if err != nil {
+		return err
+	}
+
+	// ignore query cancellations
+	ctx = context.WithValue(context.Background(), connCtxKey{}, connFromContext(ctx))
+	// recreate view to propagate schema changes
+	return c.Exec(ctx, &drivers.Statement{
+		Query:    fmt.Sprintf("CREATE OR REPLACE VIEW %s AS SELECT * FROM %s.default", safeSQLName(tableName), safeSQLName(dbName)),
+		Priority: 1,
+	})
+}
+
+// CreateTableAsSelect implements drivers.OLAPStore.
+func (c *connection) CreateTableAsSelect(ctx context.Context, name string, view bool, sql string) error {
+	if view {
+		return c.Exec(ctx, &drivers.Statement{
+			Query:       fmt.Sprintf("CREATE OR REPLACE VIEW %s AS (%s)", safeSQLName(name), sql),
+			Priority:    1,
+			LongRunning: true,
+		})
+	}
+	if !c.config.ExtTableStorage {
+		return c.Exec(ctx, &drivers.Statement{
+			Query:       fmt.Sprintf("CREATE OR REPLACE TABLE %s AS (%s)", safeSQLName(name), sql),
+			Priority:    1,
+			LongRunning: true,
+		})
+	}
+	// create a new db file in /<instanceid>/<name> directory
+	sourceDir := filepath.Join(c.config.ExtStoragePath, name)
+	if err := os.Mkdir(sourceDir, fs.ModePerm); err != nil && !errors.Is(err, fs.ErrExist) {
+		return err
+	}
+
+	// check if some older version existed previously to detach it later
+	oldVersion, oldVersionExists, _ := c.tableVersion(name)
+
+	newVersion := fmt.Sprint(time.Now().UnixMilli())
+	dbFile := filepath.Join(sourceDir, fmt.Sprintf("%s.db", newVersion))
+	db := dbName(name, newVersion)
+	// attach new db
+	err := c.Exec(ctx, &drivers.Statement{Query: fmt.Sprintf("ATTACH %s AS %s", safeSQLString(dbFile), safeSQLName(db))})
+	if err != nil {
+		removeDBFile(dbFile)
+		return err
+	}
+
+	if err := c.Exec(ctx, &drivers.Statement{
+		Query:       fmt.Sprintf("CREATE OR REPLACE TABLE %s.default AS (%s)", safeSQLName(db), sql),
+		Priority:    1,
+		LongRunning: true,
+	}); err != nil {
+		c.detachAndRemoveFile(ctx, db, dbFile)
+		return err
+	}
+
+	return c.WithConnection(ctx, 1, true, false, func(ctx, ensuredCtx context.Context, _ *dbsql.Conn) error {
+		// create view query
+		err = c.Exec(ctx, &drivers.Statement{
+			Query: fmt.Sprintf("CREATE OR REPLACE VIEW %s AS SELECT * FROM %s.default", safeSQLName(name), safeSQLName(db)),
+		})
+		if err != nil {
+			c.detachAndRemoveFile(ctx, db, dbFile)
+			return err
+		}
+
+		// success update version
+		err = c.updateVersion(name, newVersion)
+		if err != nil {
+			// extreme bad luck
+			c.detachAndRemoveFile(ctx, db, dbFile)
+			return err
+		}
+
+		if oldVersionExists {
+			oldDB := dbName(name, oldVersion)
+			// ignore these errors since source has been correctly ingested and attached
+			c.detachAndRemoveFile(ctx, oldDB, filepath.Join(sourceDir, fmt.Sprintf("%s.db", oldVersion)))
+		}
+		return nil
+	})
```

Comment:
This doesn't need `WithConnection` if you move `updateVersion` up before the `CREATE OR REPLACE VIEW`.

Reply:
Actually the `detachAndRemoveFile` need to be done with `ensuredCtx`. Missed it while doing refactor.